core/src/main/java/org/apache/iceberg/deletes/Deletes.java

```diff
@@ -261,7 +258,20 @@ public void close() {
 
     @Override
     protected boolean shouldKeep(T posDelete) {
-      return CHARSEQ_COMPARATOR.compare(dataLocation, (CharSequence) FILENAME_ACCESSOR.get(posDelete)) == 0;
+      return charSeqEquals(dataLocation, (CharSequence) FILENAME_ACCESSOR.get(posDelete));
+    }
+
+    private boolean charSeqEquals(CharSequence s1, CharSequence s2) {
+      if (s1 == s2) {
+        return true;
+      }
+
+      int count = s1.length();
+      if (count != s2.length()) {
```

Comment:
also you could group all these a bit

return s1 == s2 || (s1.length() == s2.length() && s1.toString().contentEquals(s2));

Although I'd be surprised if the length check wasn't the first thing checked in the contentEquals method so you may be able to skip that. Benchmark would help

Reply:
http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/lang/String.java#l1053 < Length check

Reply:
I'm trying to avoid the cost of `toString`. It may be negligible for a string, but not for another implementation of CharSeq.